File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/PushLocationEventTaskTest.java
package com.clevertap.android.geofence;

import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import android.content.Context;
import android.location.Location;
import com.clevertap.android.geofence.interfaces.CTGeofenceTask;
import com.clevertap.android.sdk.CleverTapAPI;
import com.google.android.gms.location.LocationResult;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import org.junit.*;
import org.mockito.*;

public class PushLocationEventTaskTest extends BaseTestCase {

    @Mock
    public CleverTapAPI cleverTapAPI;

    @Mock
    public CTGeofenceAPI ctGeofenceAPI;

    @Mock
    public CTGeofenceTask.OnCompleteListener onCompleteListener;

    private MockedStatic<CTGeofenceAPI> ctGeofenceAPIMockedStatic;

    private LocationResult locationResult;

    @Mock
    private Logger logger;

    private MockedStatic<Utils> utilsMockedStatic;

    @After
    public void cleanup() {
        ctGeofenceAPIMockedStatic.close();
        utilsMockedStatic.close();
    }

    @Before
    public void setUp() throws Exception {

        MockitoAnnotations.openMocks(this);
        super.setUp();
        Location location = new Location("");
        locationResult = LocationResult.create(Arrays.asList(new Location[]{location}));
        ctGeofenceAPIMockedStatic = mockStatic(CTGeofenceAPI.class);
        utilsMockedStatic = mockStatic(Utils.class);
        when(CTGeofenceAPI.getInstance(application)).thenReturn(ctGeofenceAPI);
        when(CTGeofenceAPI.getLogger()).thenReturn(logger);
    }

    @Test
    public void testExecuteWhenCleverTapApiIsNull() {

        when(Utils.initCTGeofenceApiIfRequired(application)).thenReturn(false);
        PushLocationEventTask task = new PushLocationEventTask(application, locationResult);

        // when listener null
        task.execute();
        Mockito.verifyNoMoreInteractions(onCompleteListener);

        // when listener not null
        task.setOnCompleteListener(onCompleteListener);
        task.execute();

        utilsMockedStatic.verify(() -> Utils.notifyLocationUpdates(any(Context.class), any(Location.class)),
                times(0));
        Mockito.verify(onCompleteListener).onComplete();
    }

    @Test
    public void testExecuteWhenCleverTapApiNotNullAndFutureIsNullAndListenerNotNull() {

        PushLocationEventTask task = new PushLocationEventTask(application, locationResult);

        // when listener not null
        task.setOnCompleteListener(onCompleteListener);
        when(Utils.initCTGeofenceApiIfRequired(application)).thenReturn(true);

        Mockito.when(cleverTapAPI.setLocationForGeofences(any(Location.class), anyInt())).
                thenReturn(null);

        task.execute();

        utilsMockedStatic.verify(() -> Utils.notifyLocationUpdates(any(Context.class), any(Location.class)));
        Mockito.verify(logger).verbose(CTGeofenceAPI.GEOFENCE_LOG_TAG,
                "Dropping location ping event to CT server");
        Mockito.verify(onCompleteListener).onComplete();
    }

    @Test
    public void testExecuteWhenCleverTapApiNotNullAndFutureIsNullAndListenerNull() {

        PushLocationEventTask task = new PushLocationEventTask(application, locationResult);

        // when listener null
        when(Utils.initCTGeofenceApiIfRequired(application)).thenReturn(true);

        Mockito.when(cleverTapAPI.setLocationForGeofences(any(Location.class), anyInt())).
                thenReturn(null);

        task.execute();

        utilsMockedStatic.verify(() -> Utils.notifyLocationUpdates(any(Context.class), any(Location.class)));
        Mockito.verify(logger).verbose(CTGeofenceAPI.GEOFENCE_LOG_TAG,
                "Dropping location ping event to CT server");
        Mockito.verifyNoMoreInteractions(onCompleteListener);
    }

    @Test
    public void testExecuteWhenCleverTapApiNotNullAndFutureNotNullAndListenerNotNull() {
        Future future = Mockito.mock(Future.class);

        PushLocationEventTask task = new PushLocationEventTask(application, locationResult);

        // when listener not null
        task.setOnCompleteListener(onCompleteListener);
        when(Utils.initCTGeofenceApiIfRequired(application)).thenReturn(true);

        Mockito.when(ctGeofenceAPI.processTriggeredLocation(any(Location.class))).
                thenReturn(future);
        task.execute();

        utilsMockedStatic.verify(() -> Utils.notifyLocationUpdates(any(Context.class), any(Location.class)));
        try {
            Mockito.verify(future).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }

        Mockito.verify(logger).verbose(CTGeofenceAPI.GEOFENCE_LOG_TAG,
                "Calling future for setLocationForGeofences()");
        Mockito.verify(onCompleteListener).onComplete();
    }

    @Test
    public void testExecuteWhenCleverTapApiNotNullAndFutureNotNullAndListenerNull() {
        Future future = Mockito.mock(Future.class);

        PushLocationEventTask task = new PushLocationEventTask(application, locationResult);

        // when listener null
        when(Utils.initCTGeofenceApiIfRequired(application)).thenReturn(true);

        Mockito.when(ctGeofenceAPI.processTriggeredLocation(any(Location.class))).
                thenReturn(future);
        task.execute();

        utilsMockedStatic.verify(() -> Utils.notifyLocationUpdates(any(Context.class), any(Location.class)));
        try {
            Mockito.verify(future).get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }

        Mockito.verify(logger).verbose(CTGeofenceAPI.GEOFENCE_LOG_TAG,
                "Calling future for setLocationForGeofences()");
        Mockito.verifyNoMoreInteractions(onCompleteListener);
    }
}

File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/PushLocationEventTaskTest.kt
package com.clevertap.android.geofence

import android.location.Location
import com.clevertap.android.geofence.interfaces.CTGeofenceTask
import com.clevertap.android.sdk.CleverTapAPI
import com.google.android.gms.location.LocationResult
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.unmockkStatic
import io.mockk.verify
import org.junit.After
import org.junit.Test
import java.util.concurrent.Future

class PushLocationEventTaskTest : BaseTestCase() {

    private lateinit var cleverTapAPI: CleverTapAPI
    private lateinit var ctGeofenceAPI: CTGeofenceAPI
    private lateinit var onCompleteListener: CTGeofenceTask.OnCompleteListener
    private lateinit var locationResult: LocationResult
    private lateinit var logger: Logger

    override fun setUp() {
        super.setUp()

        cleverTapAPI = mockk(relaxed = true)
        ctGeofenceAPI = mockk(relaxed = true)
        onCompleteListener = mockk(relaxed = true)
        logger = mockk(relaxed = true)

        locationResult = LocationResult.create(listOf(Location("")))

        mockkStatic(CTGeofenceAPI::class)
        mockkStatic(Utils::class)

        every { CTGeofenceAPI.getInstance(application) } returns ctGeofenceAPI
        every { CTGeofenceAPI.getLogger() } returns logger
    }

    @After
    override fun cleanUp() {
        super.cleanUp()
        unmockkStatic(CTGeofenceAPI::class)
        unmockkStatic(Utils::class)
    }

    @Test
    fun testExecuteWhenCleverTapApiIsNull() {
        every { Utils.initCTGeofenceApiIfRequired(application) } returns false
        val task = PushLocationEventTask(application, locationResult)

        // when listener null
        task.execute()
        confirmVerified(onCompleteListener)

        // when listener not null
        task.setOnCompleteListener(onCompleteListener)
        task.execute()

        verify(exactly = 0) { Utils.notifyLocationUpdates(any(), any()) }
        verify { onCompleteListener.onComplete() }
    }

    @Test
    fun testExecuteWhenCleverTapApiNotNullAndFutureIsNullAndListenerNotNull() {
        val task = PushLocationEventTask(application, locationResult)

        // when listener not null
        task.setOnCompleteListener(onCompleteListener)
        every { Utils.initCTGeofenceApiIfRequired(application) } returns true
        every { cleverTapAPI.setLocationForGeofences(any(), any()) } returns null
        every { ctGeofenceAPI.processTriggeredLocation(any()) } returns null

        task.execute()

        verify { Utils.notifyLocationUpdates(any(), any()) }
        verify {
            logger.verbose(
                CTGeofenceAPI.GEOFENCE_LOG_TAG, "Dropping location ping event to CT server"
            )
        }
        verify { onCompleteListener.onComplete() }
    }

    @Test
    fun testExecuteWhenCleverTapApiNotNullAndFutureIsNullAndListenerNull() {
        val task = PushLocationEventTask(application, locationResult)

        // when listener null
        every { Utils.initCTGeofenceApiIfRequired(application) } returns true
        every { cleverTapAPI.setLocationForGeofences(any(), any()) } returns null
        every { ctGeofenceAPI.processTriggeredLocation(any()) } returns null

        task.execute()

        verify { Utils.notifyLocationUpdates(any(), any()) }
        verify {
            logger.verbose(
                CTGeofenceAPI.GEOFENCE_LOG_TAG, "Dropping location ping event to CT server"
            )
        }
        verify(exactly = 0) { onCompleteListener.onComplete() }
    }

    @Test
    fun testExecuteWhenCleverTapApiNotNullAndFutureNotNullAndListenerNotNull() {
        val future: Future<*> = mockk(relaxed = true)

        val task = PushLocationEventTask(application, locationResult)

        // when listener not null
        task.setOnCompleteListener(onCompleteListener)
        every { Utils.initCTGeofenceApiIfRequired(application) } returns true
        every { ctGeofenceAPI.processTriggeredLocation(any()) } returns future

        task.execute()

        verify { Utils.notifyLocationUpdates(any(), any()) }
        verify { future.get() }

        verify {
            logger.verbose(
                CTGeofenceAPI.GEOFENCE_LOG_TAG, "Calling future for setLocationForGeofences()"
            )
        }
        verify { onCompleteListener.onComplete() }
    }

    @Test
    fun testExecuteWhenCleverTapApiNotNullAndFutureNotNullAndListenerNull() {
        val future: Future<*> = mockk(relaxed = true)

        val task = PushLocationEventTask(application, locationResult)

        // when listener null
        every { Utils.initCTGeofenceApiIfRequired(application) } returns true
        every { ctGeofenceAPI.processTriggeredLocation(any()) } returns future

        task.execute()

        verify { Utils.notifyLocationUpdates(any(), any()) }
        verify { future.get() }

        verify {
            logger.verbose(
                CTGeofenceAPI.GEOFENCE_LOG_TAG, "Calling future for setLocationForGeofences()"
            )
        }
        verify(exactly = 0) { onCompleteListener.onComplete() }
    }
}
